### PR TITLE
Alyssak branch

### PR DESCRIPTION
**Pull request by alyssakelley**
I finished making the changes from the last pull request, and this is up to PEP8 guidelines. 

I do have a question on line 59 and 129 in regards to running the build/test scripts and tagging it with a version number as well as commit hash. I know we decided on doing both tags, but I was wondering if I could get a little guidance on the best way to go about this. 

Also, I don't think that I have full permissions for the MVP Studio DockerHub because whenever I try to push to it, I keep getting access denied. I think I might only have read privileges. With that being said, this script does successfully build, push, and tag to my personal docker hub account, and it runs my build and test scripts that I made to test it. I will be working on the version tag and the unit tests next.

### CODE REVIEW

**Comment by oliverdain** [comment]
Hey Alyssa. Thanks for making the changes. I'll take a look later today or tomorrow.

For future reference, the general flow is that if a reviewer requests changes you don't close the PR. Instead you make the requested changes on the same feature branch. When you're done you `git push` again and the request a re-review (you can do that by re-assigning the review to the same person even though they already have a check-mark next to their name in the GUI). The benefit of that is the discussion isn't lost and the reviewer can click "changes since my last review" to see just what changed since that last looked at it.

Not a big deal, this is fine, but lets try the "standard flow" with the next ones.

File: build_tag_push.py [hunk 1]
```diff
@@ -1,101 +1,151 @@
...
+import os
+import subprocess
+from shutil import copyfile
+import argparse
+
+# python3 build_tag_push.py namesapce app_yaml_path
+
+# User input for the docker namespace. Refer to this as args.variable
+
+parser = argparse.ArgumentParser(description='Docker namespace')
```
Comment (oliverdain) [inline build_tag_push.py]: This is supposed to be the description for your executable as a whole. So something like `An executable to build, tag and then push a docker container` or, perhaps more commonly, `Usage: python3 build_tag_push.py <docker_hub_name> <app_yml_path>`. The kind of thing you'd see when you run, for example `sed -h` or `curl -h`

File: build_tag_push.py [hunk 2]
```diff
@@ -1,101 +1,151 @@
...
+parser.add_argument('docker_hub_name', type=str, help='DockerHub namespace')
+# parser.add_argument('docker_hub_password',
+#                     type=str, help='DockerHub password')
```
Comment (oliverdain) [inline build_tag_push.py]: Don't leave commented out code in. It just clutters things up over time.

File: build_tag_push.py [hunk 3]
```diff
@@ -1,101 +1,151 @@
...
+parser.add_argument('app_yml_path', type=str,
+                    help='Repository path to MVP Studio file')
+
+args = parser.parse_args()
```
Comment (oliverdain) [inline build_tag_push.py]: typically you would do this in `main()`. The idea is to allow you to run this code without running main to unit-test or combine this script into a bigger script that does several things. If you put this in `main` you can import this script and then call methods on it but as it you can't import this script without it parsing a non-existent command line.

File: build_tag_push.py [hunk 4]
```diff
@@ -1,101 +1,151 @@
...
+    push_docker_image(data_from_app_yml['name'])
+
+
+def parse_app_file(filename):
+    '''
+    This function is using the yaml built in to parse the file
+    which can be done by searching for the key in the dictionary
+    for the yaml file.
+
+    Contents (include in the README):
```
Comment (oliverdain) [inline build_tag_push.py]: FWIW, you don't have to duplicate documentation. It's fine to say "see the README for details".

File: build_tag_push.py [hunk 5]
```diff
@@ -1,101 +1,151 @@
...
+
+def build_check_app(data_from_app_yml):
+    '''
+    This is building and testing their app. Ignoring the return code.
+    '''
+    os.chdir(os.path.dirname(args.app_yml_path))
+    subprocess.call(['bash', data_from_app_yml['build']])
+    subprocess.call(['bash', data_from_app_yml['test']])
+    # QUESTION: I know you guys mentioned needing to ignore any failures
+    # running their test cases might cause, but how should I do this?
```
Comment (oliverdain) [inline build_tag_push.py]: I'd add a command line argument (an optional, non-positional argument) that is like `--ignore_test_errors`. Then this method takes a `ignore_test_errors` boolean and the `subprocess.call(['bash', data_from_app_yml['test']])` line becomes:

```
result = subprocess.call(['bash', data_from_app_yml['test']])
if result != 0:
   if ignore_test_errors:
       print('ERROR: tests failed')
   else:
       throw Exception('Unit tests failed')
```

Long term you'll also want to grab stdout and stderr from the tests so you can include that in your error messages but we can defer that for now.

File: build_tag_push.py [hunk 6]
```diff
@@ -1,101 +1,151 @@
...
+
+    return dict_of_contents
+
+
+def build_check_app(data_from_app_yml):
+    '''
+    This is building and testing their app. Ignoring the return code.
+    '''
+    os.chdir(os.path.dirname(args.app_yml_path))
+    subprocess.call(['bash', data_from_app_yml['build']])
```
Comment (oliverdain) [inline build_tag_push.py:61]: change to `check_call` so this throws if the build fails

File: build_tag_push.py [hunk 7]
```diff
@@ -1,101 +1,151 @@
...
+
+
+def create_context_directory(contextArray):
+    '''
+    Creating a directory that holds the docker contexts that were listed
+    in their app.yml file. This includes their build and static files.
+    Note: This directory still needs a DockerFile in it, which I am assuming
+    we will need to provide for them to use as a base file, and that
+    would need copied into this directory too, so that is what I did via the
+    last copyfile command as the end of this function.
```
Comment (oliverdain) [inline build_tag_push.py]: Comments like these shouldn't be "I decided..". They should just say when the method does. If it's a particularly odd API you might explain the reason but readers don't expect to see your thought process in API docs.

File: build_tag_push.py [hunk 8]
```diff
@@ -1,101 +1,151 @@
...
+    '''
+    repo_root_path = os.path.dirname(args.app_yml_path)
+    dir_name = "docker_context"
+    os.makedirs(repo_root_path+'/'+dir_name, exist_ok=True)
```
Comment (oliverdain) [inline build_tag_push.py]: We don't want to put the docker context right in the repo - that just kind of clutters up the user's main workspace and it's likely the user would accidentally commit that stuff to git (lots of GUI clients "helpfully" auto-add any new files). there's 2 common ways to do this:

1. Create a temporary directory. That's my vote here.
2. There's one and only one directory called `build` that gets added to the repo root. *all* "build artifacts" (e.g. the output of webpack if your doing front-end, the `.class` files if you're doing Java, etc.) go in there.

Since people are using so many different tools and they don't all use the `build` convention I suggest using just a temporary directory. See the Python `tempfile` package.

File: build_tag_push.py [hunk 9]
```diff
@@ -1,101 +1,151 @@
...
+
+    docker_context_path = repo_root_path+'/'+dir_name
```
Comment (oliverdain) [inline build_tag_push.py]: here and elsewhere, use `os.path.join` instead of `path1 + '/' + path2`. There's two reasons for that:

1. `/` isn't the right separator on Windows
2. `os.path.join` works properly if `path1` ends in a `/`. Consider `path1 = "/foo/bar/` and `path2 = "baz"` what you want, and what `os.path.join` will give you, is `/foo/bar/baz` but your code will produce `/foo/bar//baz` (note the double `/`)

File: build_tag_push.py [hunk 10]
```diff
@@ -1,101 +1,151 @@
...
+
+    copyfile(repo_root_path+'/'+contextArray[0],
+             docker_context_path+'/'+contextArray[0])
+
+    copyfile(repo_root_path+'/'+contextArray[1],
+             docker_context_path+'/'+contextArray[1])
```
Comment (oliverdain) [inline build_tag_push.py]: The `contextArray` could have any size. It might be only 1 entry in which case this line will fail, or it might be 4 entries in which case you'll fail to copy some things. So you want to use a loop.

Also not that *directories* in the context will be more common than files but you have to handle either case. In the end it probably looks something like (untested so I'm sure there's at least a typo):

```
for file_or_dir in contextArray:
  full_path = os.path.join(repo_root_path, file_or_dir)
   if os.path.is_file(full_path):
       copyfile(full_path, os.path.join(docker_context_path, file_or_dir))
    else:
        shutil.copytree(full_path, os.path.join(docker_context_path, file_or_dir))
```

File: build_tag_push.py [hunk 11]
```diff
@@ -1,101 +1,151 @@
...
+    This is building and testing their app. Ignoring the return code.
+    '''
+    os.chdir(os.path.dirname(args.app_yml_path))
+    subprocess.call(['bash', data_from_app_yml['build']])
+    subprocess.call(['bash', data_from_app_yml['test']])
+    # QUESTION: I know you guys mentioned needing to ignore any failures
+    # running their test cases might cause, but how should I do this?
+
+
+def create_context_directory(contextArray):
```
Comment (oliverdain) [inline build_tag_push.py]: remember Python variables have names like `context_array`, not `contextArray` (by convention/pep8)

File: build_tag_push.py [hunk 12]
```diff
@@ -1,101 +1,151 @@
...
+    decoded via ASCII due to the output being in bytes.
+
+    The source for this command is:
+    https://container-solutions.com/tagging-docker-images-the-right-way
+    '''
+
+    subprocess.check_call(["docker", "tag", name_from_app_yml,
+                           args.docker_hub_name+'/'+name_from_app_yml])
+
+    current_hash = subprocess.check_output(["git", "log", "-1",
```
Comment (oliverdain) [inline build_tag_push.py]: note that you can do:

```
current_hash = subprocess.check_output(
    ["git", "log", "-1", ...])
```
to make the wrapping cleaner.

File: build_tag_push.py [hunk 13]
```diff
@@ -1,101 +1,151 @@
...
+    # print(current_hash)
+    subprocess.check_call(["docker", "tag", name_from_app_yml,
+                           args.docker_hub_name +
+                           '/'+name_from_app_yml +
+                           ':'+current_hash])
+
+    # QUESTION: Did we come up with a verdict on whether or not we are doing
+    # an additional version number tag? If so, do you have any recommendations
+    # on how I should do this? I have been researching it a bit and haven't
+    # seen anything that looks like what you are wanting.
```
Comment (oliverdain) [inline build_tag_push.py]: For now this is fine but we *might* want to add a 2nd tag. If we do you'd make a REST request to dockerhub to get a list of all the current tags. Then filter that list for ones like `vxyz` where `xyz` are numbers (e.g. `v002`), then sort those and take the last one - that's the most recent version - add 1 to it and that's your new tag.

**Review by oliverdain — CHANGES_REQUESTED** [review]
This is looking really good!

Since this is "request changes" the next steps are to make the requested changes (or ask for clarification or suggest that maybe what you had was better than what I suggested, etc.) in this same branch. Then push the branch again - that will update this PR. Then you re-click me as a reviewer (I know - terrible UI on github's part) and I'll get an email that it's ready for re-review. At that point I'll be able to see what *changed* so I don't have to re-review the stuff that already looks good (which is most of it).

Nice job!

File: build_tag_push.py (hunk 4 above)
Comment (alyssakelley) [reply build_tag_push.py]: Good to know, thank you.

File: build_tag_push.py (hunk 10 above)
Comment (alyssakelley) [reply build_tag_push.py]: I definitly should have thought about that, thank you!

File: build_tag_push.py (hunk 8 above)
Comment (alyssakelley) [reply build_tag_push.py]: For clarification, where should I have this temporary file created? Would this be created on the desktop on the machine that is running the script? Then adding the docker context to that temporary file on the desktop?

Comment (alyssakelley) [reply build_tag_push.py]: Well maybe I don't need to even ask that... is using something like 
dirpath = tempfile.mkdtemp() what you had in mind?

Here it is implemented:
------------------------------------------------------------------------- 

    def create_context_directory(context_array, args_app_yml_path):
        '''
        Creating a directory that holds the docker contexts that were listed
        in their app.yml file.
        '''
        repo_root_path = os.path.dirname(args_app_yml_path)
        dirpath = tempfile.mkdtemp()

        docker_context_path = os.path.join(repo_root_path, dirpath)

        for file_or_dir in context_array:
            full_path = os.path.join(repo_root_path, file_or_dir)
            if os.path.isfile(full_path):
                copyfile(full_path, os.path.join(docker_context_path, file_or_dir))
            else:
                shutil.copytree(full_path, os.path.join(docker_context_path, file_or_dir))

        return docker_context_path

Comment (oliverdain) [reply build_tag_push.py]: > is using something like dirpath = tempfile.mkdtemp() what you had in mind?

exactly.

Comment (oliverdain) [reply build_tag_push.py]: Oh, but this looks wrong: `docker_context_path = os.path.join(repo_root_path, dirpath)`. `mkdtemp` will return a path to temporary storage on the machine and it'll be a full path. Typically it's to a dir like `/tmp` which the computer knows to delete on reboot (hence it's temporary). If you do `os.path.join(repo_root_path, dirpath)` you're going put the temporary context dir inside their repo and that's not what we want - we just want it some random place *outside* the repo and we'll delete it when we're done with the `docker build` (and if we forget the OS will delete it for us 'cause it's a temp directory).

So, in summary, `docker_context_path` **is** just `tempfile.mkdtemp()`. If you're curious feel free to add a `print('Docker context path is: %s' % docker_context_path)` to see what directory `mkdtemp` chose. You can also `ls` that directory to see all the stuff you copied there.

File: build_tag_push.py [hunk 14]
```diff
@@ -1,101 +1,155 @@
...
+
+    return dict_of_contents
+
+
+def build_check_app(data_from_app_yml, args_app_yml_path,
+                    args_ignore_test_errors):
+    '''
+    This is building and testing their app. Ignoring the return code.
+    '''
+    path_to_build_and_test = args_app_yml_path
```
Comment (slively) [inline build_tag_push.py:56]: a log message at the top of this detailing what's about to happen would be nice
```
print(f'Starting build task: {data_from_app_yml['build']]}')
...
print(f'Starting test task: {data_from_app_yml['test']]}')
...
```

File: build_tag_push.py [hunk 15]
```diff
@@ -1,101 +1,155 @@
...
+
+    else:
+        return True
+
+
+def create_context_directory(context_array, args_app_yml_path):
+    '''
+    Creating a directory that holds the docker contexts that were listed
+    in their app.yml file.
+    '''
```
Comment (slively) [inline build_tag_push.py:78]: would be good to have a print message here like above

File: build_tag_push.py [hunk 16]
```diff
@@ -1,101 +1,155 @@
...
+
+    return docker_context_path
+
+
+def build_docker_image(data_from_app_yml, args_app_yml_path):
+    '''
+    Builds the docker image with the files that are in that directory,
+    which should be a build, static, and Dockerfile. It is then named
+    with their name from their app.yml file.
+    '''
```
Comment (slively) [inline build_tag_push.py:101]: would be good to have a print message here like above

File: build_tag_push.py [hunk 17]
```diff
@@ -1,101 +1,155 @@
...
+    docker_dir_context = create_context_directory(data_from_app_yml
+                                                  ['docker_context'],
+                                                  args_app_yml_path)
+
+    image_name = data_from_app_yml['name']
+
+    result = subprocess.check_call(["docker", "build", docker_dir_context,
```
Comment (slively) [inline build_tag_push.py]: do we need to check the result? `check_call` will throw an error if it fails.

File: build_tag_push.py [hunk 18]
```diff
@@ -1,101 +1,155 @@
...
+                                        name_from_app_yml)])
+
+    current_hash = subprocess.check_output(
+        ["git", "log", "-1", "--pretty=%h"]).strip().decode('ascii')
+
+    result = subprocess.check_call(
+        ["docker", "tag", name_from_app_yml, args_docker_hub_name +
+         '/'+name_from_app_yml + ':'+current_hash])
+
+    if result != 0:
```
Comment (slively) [inline build_tag_push.py]: do we need to check the result? check_call will throw an error if it fails.

File: build_tag_push.py [hunk 19]
```diff
@@ -1,101 +1,155 @@
...
+
+
+def push_docker_image(name_from_app_yml, args_docker_hub_name):
+    '''
+    This function is to push the docker image once it has been built and
+    tagged with the name, and the commit hash tag, which is dependent on the
+    functions build_DockerImage and tag_DockerImage functions.
+    '''
+    result = subprocess.check_call(["docker", "push", args_docker_hub_name +
+                                    '/'+name_from_app_yml])
```
Comment (slively) [inline build_tag_push.py]: do we need to check the result? check_call will throw an error if it fails.

File: build_tag_push.py [hunk 20]
```diff
@@ -1,101 +1,155 @@
...
+    subprocess.check_call(['bash', data_from_app_yml['build']])
+
+    result = subprocess.call(['bash', data_from_app_yml['test']])
+
+    if result != 0:
+        if args_ignore_test_errors:
+            print("ERROR: Tests failes.")
+        else:
+            Assert.ThrowsException('Unit tests failed.')
+        return False
```
Comment (slively) [inline build_tag_push.py]: It doesn't look like we need to return any values from this function.

File: build_tag_push.py [hunk 21]
```diff
@@ -1,101 +1,155 @@
...
+    os.chdir(os.path.dirname(path_to_build_and_test))
+    subprocess.check_call(['bash', data_from_app_yml['build']])
+
+    result = subprocess.call(['bash', data_from_app_yml['test']])
+
+    if result != 0:
+        if args_ignore_test_errors:
+            print("ERROR: Tests failes.")
+        else:
+            Assert.ThrowsException('Unit tests failed.')
```
Comment (oliverdain) [inline build_tag_push.py]: I don't know what `Assert` is - I don't see anything imported and I don't see a class called `Assert` that you write. Does this work?

The standard way to throw in Python would just be `raise RuntimeError('Unit tests failed.')`

File: build_tag_push.py (hunk 18 above)
Comment (oliverdain) [reply build_tag_push.py]: +1

`subprocess.call` returns the return code and you check it. We use that for unit tests because we don't always want to throw if the test fail. OTOH `check_call` will throw if there's any runtime error so you don't have to check anything.

**Review by oliverdain — APPROVED** [review]
This looks really good to me. Made one note about the way you're trowing the exception. Other than that, will let @slively finish up the review and give the final approve.

**Review by slively — APPROVED** [review]

File: build_tag_push.py [hunk 22]
```diff
@@ -1,101 +1,147 @@
...
+import sys
+import os
+import subprocess
+from shutil import copyfile
+import argparse
+import tempfile
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Usage: python3'
```
Comment (slively) [inline build_tag_push.py:11]: this is the thing to make sure it looks good when printed out
